File: chains/rust/quantumvm/src/store.rs
// SPDX-License-Identifier: BSD-3-Clause-Eco
// Copyright (C) 2019-2026, Lux Industries Inc. All rights reserved.

//! What the chain keeps, and how a block's writes reach the device as ONE
//! thing.
//!
//! A block, its height index entry and the tip pointer move together or not at
//! all. Without that a node restarts holding a tip pointer to a block it did
//! not store, or a height index naming a block that is not there — and both are
//! states no rule can recover from, because the chain's own record of where it
//! is disagrees with what it holds.
//!
//! So the interface is not "put a key": it is [`Store::commit`], which takes
//! the whole batch. There is no way to write half of one.
//!
//! Two implementations, one meaning. [`Memory`] is for tests. [`Log`] is a
//! node's: an append-only file of committed batches, each one length-prefixed
//! and digested, replayed into an index at open. A torn tail — a batch that was
//! being written when the power went — fails its digest and is dropped, which
//! is the same all-or-nothing rule the commit gives, extended across a crash.

use std::collections::BTreeMap;
use std::fs::{File, OpenOptions};
use std::io::{Read, Seek, SeekFrom, Write};
use std::path::Path;
use std::sync::{Mutex, RwLock};

use sha2::{Digest, Sha256};

use crate::error::{Error, Result};

/// One key and the value it takes.
pub type Entry = (Vec<u8>, Vec<u8>);

/// What the chain reads from and commits to.
pub trait Store: Send + Sync {
    /// The value under `key`, or [`Error::NotFound`].
    ///
    /// A missing key and a store that could not answer are DIFFERENT facts, and
    /// this reports them differently. Answering "missing" for both is how one
    /// transient failure looks like a chain that never started.
    fn get(&self, key: &[u8]) -> Result<Vec<u8>>;

    /// Whether a key is there.
    fn has(&self, key: &[u8]) -> Result<bool> {
        match self.get(key) {
            Ok(_) => Ok(true),
            Err(Error::NotFound) => Ok(false),
            Err(e) => Err(e),
        }
    }

    /// Apply every write, or none of them, and be durable before returning.
    fn commit(&self, writes: &[Entry]) -> Result<()>;
}

/// A store in memory. Durable across nothing, which is why a node does not use
/// one.
#[derive(Default)]
pub struct Memory {
    map: Mutex<BTreeMap<Vec<u8>, Vec<u8>>>,
}

impl Memory {
    pub fn new() -> Memory {
        Memory::default()
    }

    /// How many keys are held. For tests that count what a commit wrote.
    pub fn len(&self) -> usize {
        self.map.lock().expect("store").len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

impl Store for Memory {
    fn get(&self, key: &[u8]) -> Result<Vec<u8>> {
        self.map
            .lock()
            .expect("store")
            .get(key)
            .cloned()
            .ok_or(Error::NotFound)
    }

    fn commit(&self, writes: &[Entry]) -> Result<()> {
        let mut map = self.map.lock().expect("store");
        for (k, v) in writes {
            map.insert(k.clone(), v.clone());
        }
        Ok(())
    }
}

/// One committed batch on disk:
///
/// ```text
/// [8-byte BE payload length][32-byte SHA-256 of the payload][payload]
/// ```
///
/// and the payload is `[4-byte BE key length][key][4-byte BE value length][value]`
/// repeated. Big-endian because that is the byte order this estate writes
/// lengths in, and a digest rather than a checksum because the same hash names
/// every block already.
const LEN: usize = 8;
const DIGEST: usize = 32;
const FRAME: usize = LEN + DIGEST;

/// A node's store: an append-only log of committed batches, indexed in memory.
pub struct Log {
    file: Mutex<File>,
    map: RwLock<BTreeMap<Vec<u8>, Vec<u8>>>,
}

impl Log {
    /// Open the log at `path`, replaying what is already there.
    ///
    /// A torn final batch is dropped and the file cut back to the last whole
    /// one, so the next append lands on a boundary. Under the write-then-apply
    /// order a torn batch is a batch no reader was ever told about, which is
    /// why dropping it is the correct repair rather than a loss.
    pub fn open(path: impl AsRef<Path>) -> Result<Log> {
        let path = path.as_ref();
        if let Some(dir) = path.parent() {
            if !dir.as_os_str().is_empty() {
                std::fs::create_dir_all(dir).map_err(|e| Error::Store(e.to_string()))?;
            }
        }
        let mut file = OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .truncate(false)
            .open(path)
            .map_err(|e| Error::Store(e.to_string()))?;

        let mut raw = Vec::new();
        file.read_to_end(&mut raw)
            .map_err(|e| Error::Store(e.to_string()))?;

        let mut map = BTreeMap::new();
        let mut good = 0usize;
        let mut at = 0usize;
        while at + FRAME <= raw.len() {
            let len = u64::from_be_bytes(raw[at..at + LEN].try_into().unwrap()) as usize;
            let body = at + FRAME;
            if body + len > raw.len() {
                break; // the batch was being written when the power went
            }
            let want = &raw[at + LEN..at + FRAME];
            let mut h = Sha256::new();
            h.update(&raw[body..body + len]);
            if h.finalize().as_slice() != want {
                break; // the bytes reached the device and the batch did not
            }
            for (k, v) in unpack(&raw[body..body + len])? {
                map.insert(k, v);
            }
            at = body + len;
            good = at;
        }
        if good != raw.len() {
            file.set_len(good as u64)
                .map_err(|e| Error::Store(e.to_string()))?;
        }
        file.seek(SeekFrom::End(0))
            .map_err(|e| Error::Store(e.to_string()))?;

        Ok(Log {
            file: Mutex::new(file),
            map: RwLock::new(map),
        })
    }

    /// How many committed batches the file holds.
    pub fn batches(&self) -> Result<usize> {
        let mut file = self.file.lock().expect("store");
        let at = file
            .stream_position()
            .map_err(|e| Error::Store(e.to_string()))?;
        file.seek(SeekFrom::Start(0))
            .map_err(|e| Error::Store(e.to_string()))?;
        let mut raw = Vec::new();
        file.read_to_end(&mut raw)
            .map_err(|e| Error::Store(e.to_string()))?;
        file.seek(SeekFrom::Start(at))
            .map_err(|e| Error::Store(e.to_string()))?;

        let mut n = 0;
        let mut cursor = 0usize;
        while cursor + FRAME <= raw.len() {
            let len = u64::from_be_bytes(raw[cursor..cursor + LEN].try_into().unwrap()) as usize;
            if cursor + FRAME + len > raw.len() {
                break;
            }
            cursor += FRAME + len;
            n += 1;
        }
        Ok(n)
    }
}

impl Store for Log {
    fn get(&self, key: &[u8]) -> Result<Vec<u8>> {
        self.map
            .read()
            .expect("store")
            .get(key)
            .cloned()
            .ok_or(Error::NotFound)
    }

    fn commit(&self, writes: &[Entry]) -> Result<()> {
        if writes.is_empty() {
            return Ok(());
        }
        let payload = pack(writes);
        let mut record = Vec::with_capacity(FRAME + payload.len());
        record.extend_from_slice(&(payload.len() as u64).to_be_bytes());
        let mut h = Sha256::new();
        h.update(&payload);
        record.extend_from_slice(&h.finalize());
        record.extend_from_slice(&payload);

        // Durable FIRST, visible second. A batch that is readable and not on
        // the device is a batch a restart forgets, and the chain would come
        // back naming a tip it cannot produce.
        {
            let mut file = self.file.lock().expect("store");
            let at = file
                .stream_position()
                .map_err(|e| Error::Store(e.to_string()))?;
            if let Err(e) = file.write_all(&record).and_then(|()| file.sync_all()) {
                // Cut the partial record back off, so the log stays a run of
                // whole batches even before the replay that would drop it.
                let _ = file.set_len(at);
                let _ = file.seek(SeekFrom::End(0));
                return Err(Error::Store(e.to_string()));
            }
        }

        let mut map = self.map.write().expect("store");
        for (k, v) in writes {
            map.insert(k.clone(), v.clone());
        }
        Ok(())
    }
}

fn pack(writes: &[Entry]) -> Vec<u8> {
    let mut out = Vec::new();
    for (k, v) in writes {
        out.extend_from_slice(&(k.len() as u32).to_be_bytes());
        out.extend_from_slice(k);
        out.extend_from_slice(&(v.len() as u32).to_be_bytes());
        out.extend_from_slice(v);
    }
    out
}

fn unpack(raw: &[u8]) -> Result<Vec<Entry>> {
    let mut out = Vec::new();
    let mut at = 0usize;
    while at < raw.len() {
        if at + 4 > raw.len() {
            return Err(Error::Store("a batch ends inside a key length".into()));
        }
        let klen = u32::from_be_bytes(raw[at..at + 4].try_into().unwrap()) as usize;
        at += 4;
        if at + klen + 4 > raw.len() {
            return Err(Error::Store("a batch ends inside a key".into()));
        }
        let key = raw[at..at + klen].to_vec();
        at += klen;
        let vlen = u32::from_be_bytes(raw[at..at + 4].try_into().unwrap()) as usize;
        at += 4;
        if at + vlen > raw.len() {
            return Err(Error::Store("a batch ends inside a value".into()));
        }
        let value = raw[at..at + vlen].to_vec();
        at += vlen;
        out.push((key, value));
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn temp(name: &str) -> std::path::PathBuf {
        let mut p = std::env::temp_dir();
        p.push(format!(
            "quantumvm-store-{}-{}-{:?}",
            name,
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_nanos()
        ));
        p
    }

    fn store_round_trip(s: &dyn Store) {
        assert!(matches!(s.get(b"nothing"), Err(Error::NotFound)));
        assert!(!s.has(b"nothing").unwrap());
        s.commit(&[
            (b"a".to_vec(), b"1".to_vec()),
            (b"b".to_vec(), b"2".to_vec()),
        ])
        .unwrap();
        assert_eq!(s.get(b"a").unwrap(), b"1");
        assert_eq!(s.get(b"b").unwrap(), b"2");
        assert!(s.has(b"a").unwrap());
        s.commit(&[(b"a".to_vec(), b"3".to_vec())]).unwrap();
        assert_eq!(s.get(b"a").unwrap(), b"3");
    }

    #[test]
    fn both_stores_mean_the_same_thing() {
        store_round_trip(&Memory::new());
        let path = temp("round-trip");
        store_round_trip(&Log::open(&path).unwrap());
        let _ = std::fs::remove_file(path);
    }

    #[test]
    fn what_was_committed_is_there_after_a_restart() {
        let path = temp("restart");
        {
            let log = Log::open(&path).unwrap();
            log.commit(&[
                (b"lastAccepted".to_vec(), vec![7; 32]),
                (b"height".to_vec(), 42u64.to_be_bytes().to_vec()),
            ])
            .unwrap();
        }
        let reopened = Log::open(&path).unwrap();
        assert_eq!(reopened.get(b"lastAccepted").unwrap(), vec![7; 32]);
        assert_eq!(
            reopened.get(b"height").unwrap(),
            42u64.to_be_bytes().to_vec()
        );
        assert_eq!(reopened.batches().unwrap(), 1);
        let _ = std::fs::remove_file(path);
    }

    #[test]
    fn a_batch_that_was_being_written_when_the_power_went_is_dropped_whole() {
        let path = temp("torn");
        {
            let log = Log::open(&path).unwrap();
            log.commit(&[(b"first".to_vec(), b"one".to_vec())]).unwrap();
            log.commit(&[
                (b"second".to_vec(), b"two".to_vec()),
                (b"tip".to_vec(), b"moved".to_vec()),
            ])
            .unwrap();
        }
        // Tear the last batch in half — the state a crash mid-write leaves.
        let mut raw = std::fs::read(&path).unwrap();
        let cut = raw.len() - 6;
        raw.truncate(cut);
        std::fs::write(&path, &raw).unwrap();

        let reopened = Log::open(&path).unwrap();
        assert_eq!(reopened.get(b"first").unwrap(), b"one");
        // Neither half of the torn batch survived: all or nothing holds across
        // the crash, not just across the call.
        assert!(matches!(reopened.get(b"second"), Err(Error::NotFound)));
        assert!(matches!(reopened.get(b"tip"), Err(Error::NotFound)));
        assert_eq!(reopened.batches().unwrap(), 1);

        // And the file was cut back, so the next append lands on a boundary.
        reopened
            .commit(&[(b"third".to_vec(), b"three".to_vec())])
            .unwrap();
        let again = Log::open(&path).unwrap();
        assert_eq!(again.get(b"third").unwrap(), b"three");
        assert_eq!(again.batches().unwrap(), 2);
        let _ = std::fs::remove_file(path);
    }

    #[test]
    fn a_batch_whose_bytes_were_altered_is_dropped() {
        let path = temp("altered");
        {
            let log = Log::open(&path).unwrap();
            log.commit(&[(b"key".to_vec(), b"honest".to_vec())])
                .unwrap();
        }
        let mut raw = std::fs::read(&path).unwrap();
        let last = raw.len() - 1;
        raw[last] ^= 0xFF;
        std::fs::write(&path, &raw).unwrap();

        let reopened = Log::open(&path).unwrap();
        assert!(matches!(reopened.get(b"key"), Err(Error::NotFound)));
        let _ = std::fs::remove_file(path);
    }

    #[test]
    fn an_empty_commit_writes_nothing() {
        let path = temp("empty");
        let log = Log::open(&path).unwrap();
        log.commit(&[]).unwrap();
        assert_eq!(log.batches().unwrap(), 0);
        let _ = std::fs::remove_file(path);
    }
}

File: chains/cpp/quantumvm/src/store.cpp
// Copyright (C) 2026, Lux Industries, Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause-Eco

#include "lux/quantumvm/store.hpp"

#include <zap/zap.hpp>

#include <fcntl.h>
#include <unistd.h>

#include <cerrno>
#include <cstring>
#include <vector>

namespace lux::quantumvm::store {
namespace {

// The log record: five sections, all packed lists over one blob.
constexpr std::int64_t kOffOps = 0;       // bytes: one per row, 1 = put, 0 = delete
constexpr std::int64_t kOffKeyLens = 8;   // u32 list
constexpr std::int64_t kOffKeyBlob = 16;  // bytes
constexpr std::int64_t kOffValLens = 24;  // u32 list
constexpr std::int64_t kOffValBlob = 32;  // bytes
constexpr std::int64_t kRecordSize = 40;

// The log is rewritten when it holds this much more than the rows need. Two is
// the usual amortization: every row is rewritten at most once per doubling.
constexpr std::size_t kCompactRatio = 2;
constexpr std::size_t kCompactFloor = 64 * 1024;

Bytes to_bytes(ByteView v) { return Bytes(v.begin(), v.end()); }

std::int64_t write_u32_list(zap::Builder& b, const std::vector<std::uint32_t>& xs) {
    auto lb = b.start_list(4);
    for (std::uint32_t x : xs) lb.add_u32(x);
    return lb.finish().first;
}

std::vector<std::uint32_t> read_u32_list(const zap::Object& o, std::int64_t field) {
    const zap::List l = o.list_stride(field, 4);
    std::vector<std::uint32_t> out(static_cast<std::size_t>(l.size() < 0 ? 0 : l.size()));
    for (int i = 0; i < l.size(); ++i) out[static_cast<std::size_t>(i)] = l.u32(i);
    return out;
}

Status errno_fail(Err code, const std::string& what) {
    return fail(code, what + ": " + std::strerror(errno));
}

}  // namespace

Bytes encode_batch(const Batch& batch) {
    Bytes ops, key_blob, val_blob;
    std::vector<std::uint32_t> key_lens, val_lens;
    ops.reserve(batch.size());
    key_lens.reserve(batch.size());
    val_lens.reserve(batch.size());

    for (const auto& [key, val] : batch) {
        ops.push_back(val.has_value() ? 1 : 0);
        key_lens.push_back(static_cast<std::uint32_t>(key.size()));
        key_blob.insert(key_blob.end(), key.begin(), key.end());
        val_lens.push_back(static_cast<std::uint32_t>(val ? val->size() : 0));
        if (val) val_blob.insert(val_blob.end(), val->begin(), val->end());
    }

    zap::Builder b(zap::kHeaderSize + kRecordSize + ops.size() + key_blob.size() + val_blob.size() +
                   8 * batch.size() + 64);
    const std::int64_t key_lens_off = write_u32_list(b, key_lens);
    const std::int64_t val_lens_off = write_u32_list(b, val_lens);

    auto ob = b.start_object(kRecordSize);
    ob.set_bytes(kOffOps, view(ops));
    ob.set_list(kOffKeyLens, key_lens_off, static_cast<std::int64_t>(key_lens.size()));
    ob.set_bytes(kOffKeyBlob, view(key_blob));
    ob.set_list(kOffValLens, val_lens_off, static_cast<std::int64_t>(val_lens.size()));
    ob.set_bytes(kOffValBlob, view(val_blob));
    ob.finish_as_root();
    return b.finish();
}

Result<Batch> decode_batch(ByteView record) {
    const auto msg = zap::Message::parse(record);
    if (!msg) return fail(Err::StoreCorrupt, "record does not parse");

    const zap::Object root = msg->root();
    const auto ops = root.bytes(kOffOps);
    const auto key_lens = read_u32_list(root, kOffKeyLens);
    const auto key_blob = root.bytes(kOffKeyBlob);
    const auto val_lens = read_u32_list(root, kOffValLens);
    const auto val_blob = root.bytes(kOffValBlob);

    const std::size_t n = ops.size();
    if (key_lens.size() != n || val_lens.size() != n)
        return fail(Err::StoreCorrupt, "row count mismatch");

    Batch out;
    std::size_t k_pos = 0, v_pos = 0;
    for (std::size_t i = 0; i < n; ++i) {
        const std::size_t k_len = key_lens[i];
        const std::size_t v_len = val_lens[i];
        if (k_pos + k_len > key_blob.size() || v_pos + v_len > val_blob.size())
            return fail(Err::StoreCorrupt, "row " + std::to_string(i) + " out of bounds");
        Bytes key = to_bytes(key_blob.subspan(k_pos, k_len));
        k_pos += k_len;
        if (ops[i] != 0)
            out[std::move(key)] = to_bytes(val_blob.subspan(v_pos, v_len));
        else
            out[std::move(key)] = std::nullopt;
        v_pos += v_len;
    }
    return out;
}

// ── Memory

Result<Bytes> Memory::get(ByteView key) const {
    if (closed_) return fail(Err::StoreClosed);
    auto it = rows_.find(to_bytes(key));
    if (it == rows_.end()) return fail(Err::NotFound);
    return it->second;
}

Status Memory::write(const Batch& batch) {
    if (closed_) return fail(Err::StoreClosed);
    for (const auto& [key, val] : batch) {
        if (val)
            rows_[key] = *val;
        else
            rows_.erase(key);
    }
    return ok();
}

Status Memory::close() {
    closed_ = true;
    return ok();
}

// ── File

File::~File() {
    if (fd_ >= 0) ::close(fd_);
}

Result<std::unique_ptr<File>> File::open(const std::string& path) {
    const int fd = ::open(path.c_str(), O_RDWR | O_CREAT | O_CLOEXEC, 0600);
    if (fd < 0) return fail(Err::StoreUnwritable, "cannot open " + path + ": " + std::strerror(errno));
    std::unique_ptr<File> f(new File(path, fd));
    if (auto r = f->replay(); !r) return std::unexpected(r.error());
    return f;
}

Status File::replay() {
    Bytes buf;
    const off_t end = ::lseek(fd_, 0, SEEK_END);
    if (end < 0) return errno_fail(Err::StoreUnwritable, "cannot size " + path_);
    buf.resize(static_cast<std::size_t>(end));
    if (end > 0) {
        if (::lseek(fd_, 0, SEEK_SET) < 0) return errno_fail(Err::StoreUnwritable, "cannot rewind " + path_);
        std::size_t got = 0;
        while (got < buf.size()) {
            const ssize_t n = ::read(fd_, buf.data() + got, buf.size() - got);
            if (n < 0) {
                if (errno == EINTR) continue;
                return errno_fail(Err::StoreUnwritable, "cannot read " + path_);
            }
            if (n == 0) break;
            got += static_cast<std::size_t>(n);
        }
        buf.resize(got);
    }

    std::size_t cursor = 0;
    while (cursor < buf.size()) {
        const ByteView rest(buf.data() + cursor, buf.size() - cursor);
        const auto msg = zap::Message::parse(rest);
        // A trailing record that does not parse is a commit that was
        // interrupted. It never happened, so the log is cut back to the last
        // record that did.
        if (!msg) break;
        const std::size_t len = msg->size();
        if (len == 0 || len > rest.size()) break;
        auto batch = decode_batch(rest.subspan(0, len));
        if (!batch) break;
        for (auto& [key, val] : *batch) {
            auto it = rows_.find(key);
            if (it != rows_.end()) live_bytes_ -= it->first.size() + it->second.size();
            if (val) {
                live_bytes_ += key.size() + val->size();
                rows_[key] = std::move(*val);
            } else if (it != rows_.end()) {
                rows_.erase(it);
            }
        }
        cursor += len;
    }

    log_bytes_ = cursor;
    if (cursor != buf.size() && ::ftruncate(fd_, static_cast<off_t>(cursor)) != 0)
        return errno_fail(Err::StoreUnwritable, "cannot truncate " + path_);
    if (::lseek(fd_, static_cast<off_t>(cursor), SEEK_SET) < 0)
        return errno_fail(Err::StoreUnwritable, "cannot seek " + path_);
    return ok();
}

Result<Bytes> File::get(ByteView key) const {
    if (fd_ < 0) return fail(Err::StoreClosed);
    auto it = rows_.find(to_bytes(key));
    if (it == rows_.end()) return fail(Err::NotFound);
    return it->second;
}

Status File::append(const Bytes& record) {
    std::size_t written = 0;
    while (written < record.size()) {
        const ssize_t n = ::write(fd_, record.data() + written, record.size() - written);
        if (n < 0) {
            if (errno == EINTR) continue;
            return errno_fail(Err::StoreUnwritable, "cannot write " + path_);
        }
        written += static_cast<std::size_t>(n);
    }
    // Durable when write() returns, or the crash story above is fiction.
    if (::fsync(fd_) != 0) return errno_fail(Err::StoreUnwritable, "cannot sync " + path_);
    log_bytes_ += record.size();
    return ok();
}

Status File::write(const Batch& batch) {
    if (fd_ < 0) return fail(Err::StoreClosed);
    if (batch.empty()) return ok();

    const Bytes record = encode_batch(batch);
    if (auto r = append(record); !r) return r;

    for (const auto& [key, val] : batch) {
        auto it = rows_.find(key);
        if (it != rows_.end()) live_bytes_ -= it->first.size() + it->second.size();
        if (val) {
            live_bytes_ += key.size() + val->size();
            rows_[key] = *val;
        } else if (it != rows_.end()) {
            rows_.erase(it);
        }
    }

    if (log_bytes_ > kCompactFloor && log_bytes_ > kCompactRatio * live_bytes_) return compact();
    return ok();
}

Status File::compact() {
    // One record holding every row, written beside the log and renamed over it.
    // The rename is the atomic point: a later open finds either the old log or
    // the whole new one, never a mixture.
    Batch all;
    for (const auto& [key, val] : rows_) all[key] = val;
    const Bytes record = encode_batch(all);

    const std::string tmp = path_ + ".compact";
    const int fd = ::open(tmp.c_str(), O_RDWR | O_CREAT | O_TRUNC | O_CLOEXEC, 0600);
    if (fd < 0) return errno_fail(Err::StoreUnwritable, "cannot open " + tmp);
    std::size_t written = 0;
    while (written < record.size()) {
        const ssize_t n = ::write(fd, record.data() + written, record.size() - written);
        if (n < 0) {
            if (errno == EINTR) continue;
            const std::string why = std::strerror(errno);
            ::close(fd);
            ::unlink(tmp.c_str());
            return fail(Err::StoreUnwritable, "cannot write " + tmp + ": " + why);
        }
        written += static_cast<std::size_t>(n);
    }
    if (::fsync(fd) != 0) {
        const std::string why = std::strerror(errno);
        ::close(fd);
        ::unlink(tmp.c_str());
        return fail(Err::StoreUnwritable, "cannot sync " + tmp + ": " + why);
    }
    ::close(fd);
    if (::rename(tmp.c_str(), path_.c_str()) != 0) {
        const std::string why = std::strerror(errno);
        ::unlink(tmp.c_str());
        return fail(Err::StoreUnwritable, "cannot rename over " + path_ + ": " + why);
    }

    ::close(fd_);
    fd_ = ::open(path_.c_str(), O_RDWR | O_CLOEXEC);
    if (fd_ < 0) return errno_fail(Err::StoreUnwritable, "cannot reopen " + path_);
    if (::lseek(fd_, 0, SEEK_END) < 0) return errno_fail(Err::StoreUnwritable, "cannot seek " + path_);
    log_bytes_ = record.size();
    return ok();
}

Status File::close() {
    if (fd_ >= 0) {
        ::close(fd_);
        fd_ = -1;
    }
    return ok();
}

// ── Version

Result<Bytes> Version::get(ByteView key) const {
    if (closed_) return fail(Err::StoreClosed);
    auto it = staged_.find(to_bytes(key));
    if (it != staged_.end()) {
        if (!it->second) return fail(Err::NotFound);
        return *it->second;
    }
    return base_->get(key);
}

Result<bool> Version::has(ByteView key) const {
    auto got = get(key);
    if (got) return true;
    if (got.error().code == Err::NotFound) return false;
    return std::unexpected(got.error());
}

Status Version::put(ByteView key, ByteView value) {
    if (closed_) return fail(Err::StoreClosed);
    staged_[to_bytes(key)] = to_bytes(value);
    return ok();
}

Status Version::del(ByteView key) {
    if (closed_) return fail(Err::StoreClosed);
    staged_[to_bytes(key)] = std::nullopt;
    return ok();
}

Status Version::commit() {
    if (closed_) return fail(Err::StoreClosed);
    if (staged_.empty()) return ok();
    if (auto r = base_->write(staged_); !r) return r;
    staged_.clear();
    return ok();
}

void Version::abort() { staged_.clear(); }

Status Version::close() {
    staged_.clear();
    closed_ = true;
    return ok();
}

}  // namespace lux::quantumvm::store
